modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4TcpChannel.java

```diff
@@ -34,15 +34,25 @@ public class Netty4TcpChannel implements TcpChannel {
     private final ListenableFuture<Void> closeContext = new ListenableFuture<>();
     private final ChannelStats stats = new ChannelStats();
     private final boolean rstOnClose;
+    private volatile Exception channelError = null;
 
     Netty4TcpChannel(Channel channel, boolean isServer, String profile, boolean rstOnClose, ChannelFuture connectFuture) {
         this.channel = channel;
         this.isServer = isServer;
         this.profile = profile;
         this.connectContext = new ListenableFuture<>();
         this.rstOnClose = rstOnClose;
-        addListener(this.channel.closeFuture(), closeContext);
         addListener(connectFuture, connectContext);
+        // The netty closeFuture is programmed to never return an exception.
+        // This listener takes close-causing exceptions reported during the
+        // channel lifetime, and reports them to the closeListener.
```

Comment:
Nit: I'd be inclined not to add comments like this, it's just repeating what the code says (and risks diverging from the code over time). But consider adding a javadoc to the `channelError` field. Potentially also an assertion that `closeFuture` is never completed exceptionally rather than simply stating it in a comment, that way the reader can see that it's really true.